Modules/SubModule_ASL/xASL_wrp_ResampleASL.m

```diff
@@ -244,15 +244,23 @@ function xASL_wrp_ResampleASL(x)
     elseif x.modules.asl.bTimeEncoded
         % Decoding of TimeEncoded data (Nifti is saved inside the function)
         ASL_im = xASL_quant_HadamardDecoding(PathASL{iSpace}, x.Q);
+        PLD_vector_size = length(unique(x.Q.Initial_PLD));
+        blockSize = x.Q.NumberEchoTimes * (PLD_vector_size - (PLD_vector_size/x.Q.TimeEncodedMatrixSize)); % Number of volumes per Hadamard block = EchoTimes*DecodedMatrixSize
         
-        % Hadamard Block size is calculated as number of TEs and the HadamardMatrixSize-1
-		blockSize = x.Q.NumberEchoTimes * (x.Q.TimeEncodedMatrixSize-1);
-		        
-        PWI = zeros(size(ASL_im,1), size(ASL_im,2), size(ASL_im,3), blockSize); % preallocate PWI
-        
-        for iBlock = 1:blockSize
-            PWI(:,:,:,iBlock) = xASL_stat_MeanNan(ASL_im(:,:,:,iBlock:blockSize:end), 4); % Averaged PWI4D across repetitions
+        if blockSize ~= size(ASL_im,4) % In case there is more than 1 block, we will average across repetitions
+			if mod(size(ASL_im,4), blockSize)
+				% Number of total volumes cannot be divided by blockSize
+				error(['Total number of volumes ' xASL_num2str(size(ASL_im,4)) ' cannot be composed of blocks of size ' xASL_num2str(blockSize)]);
+			else
+				PWI = zeros(size(ASL_im,1), size(ASL_im,2), size(ASL_im,3), blockSize); % preallocate PWI
+				for iBlock = 1:blockSize
+					PWI(:,:,:,iBlock) = xASL_stat_MeanNan(ASL_im(:,:,:,iBlock:blockSize:end), 4); % Averaged PWI4D across repetitions
+				end
+			end
+        else % if unique PLDs == decoded image size + HadBlock -> repetitions have different PLDs = we don't want to average
+            PWI = ASL_im;
         end
+
         
         % Save PWI4D
         fprintf('%s\n', PathPWI4D{iSpace});

```

Comment:
Did you test `xASL_wrp_ResampleASL` sections 9-11?
Do the SD images outputted make sense? And the wsCV reported?
How about the reported spatial CoV?
Any other temporary files that need deleting?

Reply:
Good point - this was probably not addressed and tested. Do you want to do this here, or in a new issue?

Reply:
I added this question to #1205, you can resolve this.